### ISSUE TITLE
Support for model options

### ISSUE THREAD

**Issue by simonw** [simonw/llm-mlx#3]
A useful guide to these is:
```bash
uv run mlx_lm.generate --help
```
Which outputs:
```
usage: mlx_lm.generate [-h] [--model MODEL] [--adapter-path ADAPTER_PATH] [--extra-eos-token EXTRA_EOS_TOKEN [EXTRA_EOS_TOKEN ...]] [--system-prompt SYSTEM_PROMPT] [--prompt PROMPT] [--max-tokens MAX_TOKENS] [--temp TEMP] [--top-p TOP_P] [--min-p MIN_P]
                       [--min-tokens-to-keep MIN_TOKENS_TO_KEEP] [--seed SEED] [--ignore-chat-template] [--use-default-chat-template] [--chat-template-config CHAT_TEMPLATE_CONFIG] [--verbose VERBOSE] [--max-kv-size MAX_KV_SIZE]
                       [--prompt-cache-file PROMPT_CACHE_FILE] [--kv-bits KV_BITS] [--kv-group-size KV_GROUP_SIZE] [--quantized-kv-start QUANTIZED_KV_START] [--draft-model DRAFT_MODEL] [--num-draft-tokens NUM_DRAFT_TOKENS]

LLM inference script

options:
  -h, --help            show this help message and exit
  --model MODEL         The path to the local model directory or Hugging Face repo. If no model is specified, then mlx-community/Llama-3.2-3B-Instruct-4bit is used.
  --adapter-path ADAPTER_PATH
                        Optional path for the trained adapter weights and config.
  --extra-eos-token EXTRA_EOS_TOKEN [EXTRA_EOS_TOKEN ...]
                        Add tokens in the list of eos tokens that stop generation.
  --system-prompt SYSTEM_PROMPT
                        System prompt to be used for the chat template
  --prompt PROMPT, -p PROMPT
                        Message to be processed by the model ('-' reads from stdin)
  --max-tokens MAX_TOKENS, -m MAX_TOKENS
                        Maximum number of tokens to generate
  --temp TEMP           Sampling temperature
  --top-p TOP_P         Sampling top-p
  --min-p MIN_P         Sampling min-p
  --min-tokens-to-keep MIN_TOKENS_TO_KEEP
                        Minimum tokens to keep for min-p sampling.
  --seed SEED           PRNG seed
  --ignore-chat-template
                        Use the raw prompt without the tokenizer's chat template.
  --use-default-chat-template
                        Use the default chat template
  --chat-template-config CHAT_TEMPLATE_CONFIG
                        Additional config for `apply_chat_template`. Should be a dictionary of string keys to values represented as a JSON decodable string.
  --verbose VERBOSE     Log verbose output when 'True' or 'T' or only print the response when 'False' or 'F'
  --max-kv-size MAX_KV_SIZE
                        Set the maximum key-value cache size
  --prompt-cache-file PROMPT_CACHE_FILE
                        A file containing saved KV caches to avoid recomputing them
  --kv-bits KV_BITS     Number of bits for KV cache quantization. Defaults to no quantization.
  --kv-group-size KV_GROUP_SIZE
                        Group size for KV cache quantization.
  --quantized-kv-start QUANTIZED_KV_START
                        When --kv-bits is set, start quantizing the KV cache from this step onwards.
  --draft-model DRAFT_MODEL
                        A model to be used for speculative decoding.
  --num-draft-tokens NUM_DRAFT_TOKENS
                        Number of tokens to draft when using speculative decoding.

```
Some of these should be model options, but others might be worth setting at the register/download model step.

**Comment by simonw** [comment]
I think I'll start with these:
```
  --max-tokens MAX_TOKENS, -m MAX_TOKENS
                        Maximum number of tokens to generate
  --temp TEMP           Sampling temperature
  --top-p TOP_P         Sampling top-p
  --min-p MIN_P         Sampling min-p
  --min-tokens-to-keep MIN_TOKENS_TO_KEEP
                        Minimum tokens to keep for min-p sampling.
  --seed SEED           PRNG seed
```

**Comment by simonw** [comment]
It looks like max tokens defaults to 256 if you don't specify that: https://github.com/ml-explore/mlx-examples/blob/96bf37008e91de86538bdacf3a12a479a322902b/llms/mlx_lm/utils.py#L222-L236

```python
def generate_step(
    prompt: mx.array,
    model: nn.Module,
    *,
    max_tokens: int = 256,
    sampler: Optional[Callable[mx.array, mx.array]] = None,
    logits_processors: Optional[List[Callable[[mx.array, mx.array], mx.array]]] = None,
    max_kv_size: Optional[int] = None,
    prompt_cache: Optional[Any] = None,
    prefill_step_size: int = 512,
    kv_bits: Optional[int] = None,
    kv_group_size: int = 64,
    quantized_kv_start: int = 0,
    prompt_progress_callback: Optional[Callable[int, int]] = None,
) -> Generator[Tuple[mx.array, mx.array], None, None]:
```

**Comment by simonw** [comment]
The arguments for that are passed as `**kwargs` passed to `.generate_step()`: https://github.com/ml-explore/mlx-examples/blob/96bf37008e91de86538bdacf3a12a479a322902b/llms/mlx_lm/utils.py#L222-L236

**Comment by simonw** [comment]
The other options need to be used to create a custom sampler:

https://github.com/ml-explore/mlx-examples/blob/96bf37008e91de86538bdacf3a12a479a322902b/llms/mlx_lm/generate.py#L249-L256

```python
    sampler = make_sampler(args.temp, args.top_p, args.min_p, args.min_tokens_to_keep)
    response = generate(
        model,
        tokenizer,
        prompt,
        max_tokens=args.max_tokens,
        verbose=args.verbose,
        sampler=sampler,
```

**Comment by simonw** [comment]
More relevant code:
```python
    parser.add_argument(
        "--max-tokens",
        "-m",
        type=int,
        default=DEFAULT_MAX_TOKENS,
        help="Maximum number of tokens to generate",
    )
    parser.add_argument(
        "--temp", type=float, default=DEFAULT_TEMP, help="Sampling temperature"
    )
    parser.add_argument(
        "--top-p", type=float, default=DEFAULT_TOP_P, help="Sampling top-p"
    )
    parser.add_argument(
        "--min-p", type=float, default=DEFAULT_MIN_P, help="Sampling min-p"
    )
    parser.add_argument(
        "--min-tokens-to-keep",
        type=int,
        default=DEFAULT_MIN_TOKENS_TO_KEEP,
        help="Minimum tokens to keep for min-p sampling.",
    )
    parser.add_argument("--seed", type=int, default=DEFAULT_SEED, help="PRNG seed")
```
From https://github.com/ml-explore/mlx-examples/blob/96bf37008e91de86538bdacf3a12a479a322902b/llms/mlx_lm/generate.py#L63C1-L85C84

Those default values are:
```python
DEFAULT_MAX_TOKENS = 100
DEFAULT_TEMP = 0.0
DEFAULT_TOP_P = 1.0
DEFAULT_MIN_P = 0.0
DEFAULT_MIN_TOKENS_TO_KEEP = 1
DEFAULT_SEED = 0
```

**Comment by simonw** [comment]
From digging around in the code you can set `max_tokens` to -1 for unlimited - which seems to work. But when I tried with a high temperature I got a LOT of junk!

```bash
llm -m ml 'best fact about a pelican' -o temperature 4
```
https://gist.github.com/simonw/1d166084b0aa04a7bf5204a695df70b4

I'm not sure if I should leave it at -1 as the default or if I should default to a higher value and document that you can pass `-1` to be unlimited.

**Comment by simonw** [comment]
Code so far is in this PR:
- #4

**Comment by simonw** [comment]
I could even have a `-o unlimited 1` option which makes it unlimited, which would I think be more clear than telling people to set `-o max_tokens -1`.

**Comment by simonw** [comment]
Being able to set these on a per-model basis does make some sense.

**Comment by simonw** [comment]
I've so far been unable to demonstrate that the `-o seed x` option does anything.

**Comment by simonw** [comment]
For want of a decision: I'm going to default to 1024 maximum output tokens and support a `-o unlimited 1` option for setting that to -1.

**Comment by simonw** [comment]
I couldn't figure out a clean way to have a field validator for pydantic that works in both v1 and v2 and fails if both unlimited and max_tokens are set, so I'm going to just have max_token signored if unlimited is set.

This will be easier to solve when llm fully embraces pydantic v2 and ditches v1.

**Comment by simonw** [comment]
Another decision I need to make is default values for these settings:
```python
DEFAULT_TEMPERATURE = 0.0
DEFAULT_TOP_P = 1.0
DEFAULT_MIN_P = 0.0
DEFAULT_MIN_TOKENS_TO_KEEP = 1
DEFAULT_MAX_TOKENS = 1024
```

**Comment by simonw** [comment]
I'm going to copy the defaults used by `llama.cpp`: https://github.com/ggml-org/llama.cpp/blob/68ff663a04ed92044a9937bcae353e9d9733f9cd/examples/main/README.md#temperature
```
temperature: 0.8
top-p: 0.9
min-p: 0.1
```

**Comment by simonw** [comment]
The combination of `-o seed x` and `-o temperature 1.0` appears to provide repeatable results:
```
% uv run llm -m qwen4 hi -o temperature 1.0 -o seed 1
Hello, how may I assist you today?
% uv run llm -m qwen4 hi -o temperature 1.0 -o seed 1
Hello, how may I assist you today?
% uv run llm -m qwen4 hi -o temperature 1.0 -o seed 1
Hello, how may I assist you today?
% uv run llm -m qwen4 hi -o temperature 1.0 -o seed 2
Hello! How can I assist you today? Feel free to ask me any question you have about anything you want. I'm here to help.
% uv run llm -m qwen4 hi -o temperature 1.0 -o seed 2
Hello! How can I assist you today? Feel free to ask me any question you have about anything you want. I'm here to help.
% uv run llm -m qwen4 hi -o temperature 1.0 -o seed 2
Hello! How can I assist you today? Feel free to ask me any question you have about anything you want. I'm here to help.
```

**Comment by simonw** [comment]
Documented here: https://github.com/simonw/llm-mlx/blob/b6f2ff02e3c71b5ae786f5486d23b896d836a73c/README.md#model-options

**Comment by avoidik** [comment]
is there a way to list all `-o` supported options without knowing them upfront?

**Comment by eudinaesis** [comment]
I'm probably just doing something wrong, but I can't seem to get the model options working. I followed the instructions [here](https://simonwillison.net/2025/May/2/qwen3-8b/) to set unlimited tokens (and then later max_tokens 999999), and it seemed to work:

```
llm models options
mlx-community/Qwen3-8B-4bit:
  unlimited: 1
  max_tokens: 999999
mlx-community/Qwen3-30B-A3B-4bit:
  unlimited: 1
  max_tokens: 999999
```

But when I run it without an explicit -o unlimited 1 or -o max_tokens 999999, it cuts off. When I put the option in explicitly, it does not cut off. I'm running llm 0.24.2, llm-mlx 0.4. Any ideas?
